Homework 2 rework

BrynzaA opened the pull request:
Reworked Homework#2

BrynzaA committed:
Homework 2 rework

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/controller/EmployeeController.java:
```diff
@@ -3,54 +3,32 @@
-        }
-
-        return ResponseEntity.ok(employee);
+        return ResponseEntity.ok(employeeService.getById(id));
```
а что если в запросе придёт нулл? Сервис выкинет NPE, а что увидит пользователь?

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/controller/EmployeeController.java:
```diff
@@ -3,54 +3,32 @@
-        employeeService.save(employeeDto);
-        return ResponseEntity.status(HttpStatus.CREATED).build();
+    public ResponseEntity<Employee> addEmployee(@RequestBody EmployeeDto employeeDto){
+        return ResponseEntity.status(HttpStatus.CREATED).body(employeeService.save(employeeDto));
```
аналогично, что будет если придёт нулл? Какой код ответа вернёт приложение?

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/service/Impl/StaffServiceImpl.java:
```diff
@@ -5,36 +5,47 @@
     @Override
     public Staff getById(Long id) {
-        return staffRepository.findById(id).orElse(null);
+        ValidationUtils.idIsNotNull(id);
```
Не совсем понятна логика такого решения. По факту внутри просто isNotNull и выкидывание ошибки в приложении. Что увидит пользователь если выкинется ошибка? 500 код на все ошибки?

ramilmmk commented on src/main/resources/db/migration/V1__init_DB.sql:
```diff
@@ -4,15 +4,16 @@ CREATE TABLE IF NOT EXISTS Employee
-    dateOfEmployment date         NOT NULL
+    surname          varchar(255) NOT NULL,
+    date_of_birth      date         NOT NULL,
+    date_of_employment date         NOT NULL
```
Изменения в скриптах не нужно делать. Если бы у твоих коллег с твоей команды уже выполнился этот скрипт, изменения применились бы в базе, то после твоих изменений проект просто перестал бы запускаться, так как у этого файла не совпадали бы чексуммы. В этом и суть скриптов миграции, чтобы на каждое изменение - новый скрипт с новой версией.

Korgutlova replied:
@ramilmm возможно это моя ошибка, я сказала о переименовании колонок для бд + добавленией констрейнта на FK. но лучше эти изменения было добавить отдельным скриптом

Korgutlova commented on src/main/java/com/simbirsoft/springcourse/controller/EmployeeController.java:
```diff
@@ -3,54 +3,32 @@
-
+    public void delete(@PathVariable Long id){
         employeeService.delete(id);
-        return ResponseEntity.ok().build();
```
думаю надо было оставить возвращаемый тип ResponseEntity

Korgutlova commented on src/main/java/com/simbirsoft/springcourse/controller/StaffController.java:
```diff
@@ -2,43 +2,26 @@
-
+    public void delete(@PathVariable Long id){
         staffService.delete(id);
-        return ResponseEntity.ok().build();
```
думаю надо было оставить возвращаемый тип ResponseEntity

Korgutlova commented on src/main/java/com/simbirsoft/springcourse/controller/StaffController.java:
```diff
@@ -2,43 +2,26 @@
-        return ResponseEntity.ok(staff);
+        return ResponseEntity.ok(staffService.getById(id));
     }
 
```
Почему в StaffController нет метода create?

Korgutlova reviewed (commented):
Незначительные замечания

BrynzaA committed:
Homework 2 rework
-Добавлен метод create для StaffController
-Учтены замечания относительно обработки исключений

Korgutlova reviewed (commented):
Хорошо, что логика обработки/сохранения/поиска входящих данных делегированы сервисам. Однако сервисы не должны отвечать за отправку `ResponseEntity`. Это только относится к контроллерам. 

Мы вызываем метод у сервиса, если к нас приходит валидное значение (например не null или когда не выбрасывается исключение то отправляем ok() и сущность, если что-то пошло не так, например можем отловить исключение и отправить статус с ошибкой и сообщение для пользователя, что произошла ошибка во время сохранения например). 

В целом у меня комментариев больше нет, @ramilmm что скажешь?

BrynzaA committed:
Homework 2 rework
-Добавлены новые исключения
-Добавлен обработчик исключений

Korgutlova commented:
Можно мерджить

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/service/Impl/EmployeeServiceImpl.java:
```diff
@@ -4,29 +4,50 @@
     @Override
-    public Employee getById(Long id) {
-        return employeeRepository.findById(id).orElse(null);
+    public ResponseEntity<Employee> getById(Long id) {
```
не нужно в сервисах ResponseEntity использовать. В сервисах ты должен работать с объектами своего приложения. В сервисах только бизнес-логика, т.е. какие-то операции над объектами приложения. ResponseEntity нужен для контроллера.

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/service/Impl/EmployeeServiceImpl.java:
```diff
@@ -4,29 +4,50 @@
         employee.setDateOfEmployment(employeeDto.getDateOfEmployment());
 
-        employeeRepository.save(employee);
+        return ResponseEntity.status(HttpStatus.CREATED).build();
```
респонсЕнтити тут не нужен, он должен быть в контроллерах

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/service/Impl/StaffServiceImpl.java:
```diff
@@ -5,36 +5,50 @@
+        if (isEmpty(staff)){
+            return ResponseEntity.notFound().build();
+        }
+        return ResponseEntity.ok(staff);
```
респонсЕнтити тут не нужен, он должен быть в контроллерах

ramilmmk commented on src/main/java/com/simbirsoft/springcourse/service/StaffService.java:
```diff
@@ -2,10 +2,11 @@
-    Staff getById(Long id);
-    void save(StaffDto employeeDto);
+    ResponseEntity<Staff> getById(Long id);
+    ResponseEntity<String> save(StaffDto employeeDto);
```
респонсЕнтити тут не нужен, он должен быть в контроллерах

ramilmmk reviewed (commented):
Перенеси все ResponseEntity в контроллеры, они не должны быть в сервисах
